server/src/main/java/org/elasticsearch/readiness/ReadinessService.java

```diff
@@ -0,0 +1,198 @@
+import org.elasticsearch.common.transport.PortsRange;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.shutdown.PluginShutdownService;
+import org.elasticsearch.transport.BindTransportException;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.StandardProtocolFamily;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+import static org.elasticsearch.node.Node.WRITE_PORTS_FILE_SETTING;
+import static org.elasticsearch.node.Node.writePortsFile;
+
+public class ReadinessService extends AbstractLifecycleComponent implements ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(ReadinessService.class);
+
+    private final Environment environment;
+
+    private volatile boolean active = false;
+    private volatile ServerSocketChannel serverChannel;
+    final AtomicReference<InetSocketAddress> boundSocket = new AtomicReference<>();
+
+    public static final Setting<String> PORT = new Setting<>(
+        "readiness.port",
+        "9400-9500",
+        Function.identity(),
+        Setting.Property.NodeScope
+    );
+
+    public ReadinessService(ClusterService clusterService, Environment environment) {
+        this.serverChannel = null;
+        this.environment = environment;
+        clusterService.addListener(this);
+    }
+
+    // package private for testing
+    boolean ready() {
+        return this.serverChannel != null;
+    }
+
+    // package private for testing
+    ServerSocketChannel serverChannel() {
+        return serverChannel;
+    }
+
+    public BoundTransportAddress boundAddress() {
+        TransportAddress publishAddress = new TransportAddress(boundSocket.get());
+        return new BoundTransportAddress(new TransportAddress[] { publishAddress }, publishAddress);
+    }
+
+    ServerSocketChannel setupSocket() {
+        AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+            if (boundSocket.get() != null) {
+                try {
+                    serverChannel = ServerSocketChannel.open(StandardProtocolFamily.INET);
+                    serverChannel.bind(boundSocket.get());
+                } catch (Exception e) {
+                    throw new BindTransportException("Failed to re-bind to " + boundSocket.get(), e);
+                }
+
+                return null;
+            }
+
+            final AtomicReference<Exception> lastException = new AtomicReference<>();
+
+            PortsRange portsRange = new PortsRange(PORT.get(environment.settings()));
+            InetAddress localhost = InetAddress.getLoopbackAddress();
+
+            boolean success = portsRange.iterate(portNumber -> {
```

Comment:
... moreover I think this means we might get a different socket each time we become ready. I mean today that's only once in the node's lifecycle, but we don't expect this to be true forever.

Reply:
I did add code to prevent this, just a bit above these lines. If I had already bound this process to a given port, it reattempts only on that port again and fails to bind if it can't. I can probably scrap this code if we explicitly configure the port to a single value.

Reply:
Ah yes sorry I missed that.